xonsh/environ.py

```diff
@@ -7,62 +7,188 @@
 import builtins
 import platform
 import subprocess
+from datetime import datetime
+from functools import partial
 from warnings import warn
 
 from xonsh.tools import TERM_COLORS
 
-def current_branch(cwd=None):
-    """Gets the branch for a current working directory. Returns None
-    if the cwd is not a repository.  This currently only works for git, 
-    bust should be extended in the future.
-    """
-    branch = None
-    cwd = os.getcwd() if cwd is None else cwd
+class PromptFormatter(dict):
+    '''
+    Base class that implements utilities for PromptFormatters.
 
-    # step out completely if git is not installed
-    try:
-        binary_location = subprocess.check_output(['which', 'git'], cwd=cwd,
-                                    stderr=subprocess.PIPE,
-                                    universal_newlines=True)
-        if not binary_location:
-            return branch
-    except subprocess.CalledProcessError:
-        return branch
+    You should read the docs for this class to learn how to write your own
+    Prompt Formatter.  However, you most likely want to inherit from
+    DefaultPromptFormatter if you are writing a custom formatter.
+
+    Custom Prompt Formatters will need to implement their own __init__()
+    method that adds entries to several dicts depending on the needs of the
+    variables they add.  The Formatter may also implement methods to implement
+    those variables (although very simple variables can be written with
+    pre-existing functions.)
+
+    Example:
```

Comment:
<nod>  Yeah -- about that, I was trying to build the docs so that I could check how all of this affects the docstrings.  In addition to cleaning up the example formatting there also needs to be documentation updates to add the new prompt variables.  And I'm not sure if xonsh has configured sphinx to pull documentation from the class's docstring or the ```__init___``` docstring (When I've used sphinx in the past it's been an either or).

Unfortunately, I couldn't figure out how to build the docs... I got to the point where I had to modify the Makefile to use sphinx-build-3.4 so that it could use autodoc on the source files but then it appeared that cloud_sptheme wasn't ported to python3 yet. (pip3.4 install --user cloud_sptheme fails).  Any clues on how you do it so that I can be sure the documentation is all building correctly?